### PR TITLE
Ddms mining consensus

### PR DESCRIPTION
**Pull request by PJLewandowski**

### CODE REVIEW

File: src/policy/ddms.cpp [hunk 1]
```diff
@@ -89,6 +102,177 @@ void MinerLicenses::ModifyLicense(const MinerLicenses::LicenseEntry& entry) {
...
-	else
+	else {
 		license->hashRate = entry.hashRate;
+		license->height = entry.height;
+	}
+}
+
+std::string CScriptToAddressString(const CScript& scriptPubKey) {
+	auto scriptStr = HexStr(scriptPubKey.begin(), scriptPubKey.end());
+	return scriptStr.substr(4, scriptStr.size() - 6);
```
Comment (mtwaro) [inline src/policy/ddms.cpp]: Are you sure about these values? I suppose you want to exctract hash160 from script, but this will work only for P2PKH scripts.

File: .settings/language.settings.xml [hunk 2]
```diff
@@ -0,0 +1,52 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
```
Comment (mtwaro) [inline .settings/language.settings.xml]: Is this file needed in repository?

File: src/policy/ddms.cpp [hunk 3]
```diff
@@ -36,11 +49,7 @@ std::vector<MinerLicenses::LicenseEntry> MinerLicenses::ExtractLicenseEntries(co
 MinerLicenses::LicenseEntry MinerLicenses::ExtractLicenseEntry(const CScript& scriptPubKey, const int height) {
 	const int size = scriptPubKey.size();
 	uint16_t hashRate = scriptPubKey[size - 2] << 8 | scriptPubKey[size - 1];
-	std::string address = "";
-
-	// TODO make it prettier
-	for( int i = 5; i < 5 + MinerScriptSize(scriptPubKey); ++i)
-		address += static_cast<char>(scriptPubKey[i]);
+	std::string address = HexStr(scriptPubKey.begin() + 5, scriptPubKey.begin() + 5 + MinerScriptSize(scriptPubKey));
```
Comment (mtwaro) [inline src/policy/ddms.cpp]: What is this variable? Is this a hash160 of a public key / script? If so, we should not call it an address. Could you also explain the meaning of `5`?

File: src/policy/ddms.cpp [hunk 4]
```diff
@@ -89,6 +102,177 @@ void MinerLicenses::ModifyLicense(const MinerLicenses::LicenseEntry& entry) {
...
+
+	for (auto& entry : minersBlockAverage)
+		entry.second /= rounds;
+
+	return minersBlockAverage;
+}
+
+float MiningMechanism::CalcMinerBlockAverageOnAllRounds(const CScript& scriptPubKey, const uint32_t heightThreshold) {
+	const auto scriptStrAddr = CScriptToAddressString(scriptPubKey);
+	return CalcMinersBlockAverageOnAllRounds(heightThreshold)[scriptStrAddr];
```
Comment (mtwaro) [inline src/policy/ddms.cpp]: It's quite expensive to do the full calculation on every request. Maybe the map should be stored as a variable?

File: src/policy/ddms.cpp [hunk 5]
```diff
@@ -89,6 +102,177 @@ void MinerLicenses::ModifyLicense(const MinerLicenses::LicenseEntry& entry) {
...
+float MinerLicenses::GetHashrateSum() const {
+	return std::accumulate(std::begin(licenses), std::end(licenses), 0.0f, [](float result, const MinerLicenses::LicenseEntry& license) {
+		return result + license.hashRate;
+	});
+}
+
+std::unordered_map<std::string, float> MiningMechanism::CalcMinersBlockAverageOnAllRounds(uint32_t heightThreshold) {
+	std::unordered_map<std::string, float> minersBlockAverage;
+	auto blockIndex = chainActive.Tip();
+	uint16_t rounds = blockIndex->nHeight % MINING_ROUND_SIZE != MINING_ROUND_SIZE - 1;
```
Comment (mtwaro) [inline src/policy/ddms.cpp]: For clarity, and to avoid implicit casting, I would turn this expression into:
`uint16_t rounds = blockIndex->nHeight % MINING_ROUND_SIZE != MINING_ROUND_SIZE - 1 ? 1 : 0`

File: src/policy/ddms.cpp [hunk 6]
```diff
@@ -89,6 +102,177 @@ void MinerLicenses::ModifyLicense(const MinerLicenses::LicenseEntry& entry) {
...
+	return CalcMinersBlockAverageOnAllRounds(heightThreshold)[scriptStrAddr];
+}
+
+std::unordered_map<std::string, int> MiningMechanism::CalcMinersBlockQuota() {
+	std::unordered_map<std::string, int> minersBlockQuota;
+	auto licenses = minerLicenses.GetLicenses();
+	float hashrateSum = minerLicenses.GetHashrateSum();
+
+	for (const auto& license : licenses)
+		minersBlockQuota[license.address] = std::round(MINING_ROUND_SIZE * license.hashRate / hashrateSum);
```
Comment (mtwaro) [inline src/policy/ddms.cpp]: Consider a following case:
- miner 1: 3 PH
- miner 2: 3 PH
- miner 3: 3 PH
total: 9 PH
after calculations:
- miner 1: 33 blocks
- miner 2: 33 blocks
- miner 3: 33 blocks
total: 99 blocks
1 block is left off due to the rounding error. If we have 29 miners with the same hashrates, not 1 but 13 blocks will be unaccounted for. We need to be prepared for this scenario

File: src/policy/ddms.cpp [hunk 7]
```diff
@@ -89,6 +102,177 @@ void MinerLicenses::ModifyLicense(const MinerLicenses::LicenseEntry& entry) {
...
+
+	return minersBlockQuota;
+}
+
+uint16_t MiningMechanism::CalcMinerBlockQuota(const CScript& scriptPubKey) {
```
Comment (mtwaro) [inline src/policy/ddms.cpp]: Same as in line 152. It's inelegant to recalculate everything for every user.

File: src/policy/ddms.cpp [hunk 8]
```diff
@@ -89,6 +102,177 @@ void MinerLicenses::ModifyLicense(const MinerLicenses::LicenseEntry& entry) {
...
+				--minersBlockLeftInRound[scriptStrAddr];
+		}
+
+		currentBlockIndex = currentBlockIndex->pprev;
+	}
+
+	return minersBlockLeftInRound;
+}
+
+uint16_t MiningMechanism::CalcMinerBlockLeftInRound(const CScript& scriptPubKey, const uint32_t heightThreshold) {
```
Comment (mtwaro) [inline src/policy/ddms.cpp]: Same as lines 152 and 174

File: src/policy/ddms.cpp [hunk 9]
```diff
@@ -1,10 +1,23 @@
...
 #include <primitives/transaction.h>
+#include <util/strencodings.h>
 
-const CScript WDMO_SCRIPT = CScript() << OP_HASH160 << std::vector<unsigned char>{11, 182, 127, 3, 232, 176, 211, 69, 45, 165, 222, 55, 211, 47, 198, 174, 240, 165, 160, 160} << OP_EQUAL; // TODO: replace with actual wdmo script; use ParseHex
 MinerLicenses minerLicenses{};
+MiningMechanism miningMechanism{};
+const CScript WDMO_SCRIPT = CScript() << OP_HASH160 << std::vector<unsigned char>{11, 182, 127, 3, 232, 176, 211, 69, 45, 165, 222, 55, 211, 47, 198, 174, 240, 165, 160, 160} << OP_EQUAL; // TODO: replace with actual wdmo script; use ParseHex
+const uint16_t MINING_ROUND_SIZE{ 100 };
+const uint32_t FIRST_MINING_ROUND_HEIGHT{ 35000 }; // TODO: change to proper value
+const uint32_t MAX_CLOSED_ROUND_TIME { MAX_FUTURE_BLOCK_TIME * 5 };
```
Comment (mtwaro) [inline src/policy/ddms.cpp]: Too big. I think we should put 3 times, maybe 4

**Review by mtwaro — CHANGES_REQUESTED** [review]

**Comment by mtwaro** [comment]
Also, merge conflicts have to be resolved.

**Comment by rynomster** [comment]
Good evening

Since there are no issues and no public email addresses I will write here.
We seem to be having problems with our Bitcoin Vault node, where it just stops responding every few hours then we have to SIGTERM and restart it before we can use it.

Where can we open issues or talk to the developers or anyone in the project?

File: src/policy/ddms.cpp (hunk 1 above)
Comment (PJLewandowski) [reply src/policy/ddms.cpp]: Done

File: .settings/language.settings.xml (hunk 2 above)
Comment (PJLewandowski) [reply .settings/language.settings.xml]: Done

File: src/policy/ddms.cpp (hunk 3 above)
Comment (PJLewandowski) [reply src/policy/ddms.cpp]: Done

File: src/policy/ddms.cpp (hunk 5 above)
Comment (PJLewandowski) [reply src/policy/ddms.cpp]: Done (function removed).

File: src/policy/ddms.cpp (hunk 6 above)
Comment (PJLewandowski) [reply src/policy/ddms.cpp]: It shouldnt be a problem. Left blocks will just define a minimum number of open-ring subround.

File: src/policy/ddms.cpp (hunk 7 above)
Comment (PJLewandowski) [reply src/policy/ddms.cpp]: Done

File: src/policy/ddms.cpp (hunk 8 above)
Comment (PJLewandowski) [reply src/policy/ddms.cpp]: Done

File: src/policy/ddms.cpp (hunk 9 above)
Comment (PJLewandowski) [reply src/policy/ddms.cpp]: Done

File: src/policy/ddms.cpp (hunk 4 above)
Comment (PJLewandowski) [reply src/policy/ddms.cpp]: Done (function removed)

**Comment by rynomster** [comment]
@PJLewandowski any news on the comment above? Also which version should we be using, as we are getting 76/100 blocks have unknown version

**Comment by mtwaro** [comment]
@rynomster pull request discussions are not support chats, please do not use them this way. Here is a public e-mail you can contact to get help: contact@bitcoinvault.global.

File: src/rpc/blockchain.cpp [hunk 10]
```diff
@@ -2333,6 +2371,7 @@ static const CRPCCommand commands[] =
     { "blockchain",         "getblockheader",         &getblockheader,         {"blockhash","verbose"} },
     { "blockchain",         "getchaintips",           &getchaintips,           {} },
     { "blockchain",         "getdifficulty",          &getdifficulty,          {} },
+	{ "blockchain", 		"getlicensedminers",	  &getlicensedminers,	   {} },
```
Comment (mtwaro) [inline src/rpc/blockchain.cpp]: Indentation

File: src/rpc/blockchain.cpp [hunk 11]
```diff
@@ -2319,6 +2320,43 @@ UniValue scantxoutset(const JSONRPCRequest& request)
...
+{
+    if (request.fHelp || request.params.size() != 0)
+        throw std::runtime_error(
+            RPCHelpMan{"getlicensedminers",
+                "\nReturns the list of licensed miners.\n",
+                {},
+				RPCResult{
+				            "\"licenses\": [\n"
+				            "    {\n"
+				            "      \"hashrate_ph\": n,             (numeric) Hashrate assigned to miner\n"
```
Comment (mtwaro) [inline src/rpc/blockchain.cpp]: `(numeric) Hashrate assigned to miner (in PH/s)\n`

**Review by mtwaro — APPROVED** [review]
